p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,351 @@
+	return "0.0.0"
+}
+
+func (l LazyLedgerPlugin) Init(env *plugin.Environment) error {
+	return nil
+}
+
+// DataSquareRowOrColumnRawInputParser reads the raw shares and extract the IPLD nodes from the NMT tree.
+// Note, to parse without any error the input has to be of the form:
+//
+// <share_0>| ... |<share_numOfShares>
+//
+// To determine the share and the namespace size the constants
+// types.ShareSize and types.NamespaceSize are used.
+func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int) ([]node.Node, error) {
+	const extendedSquareSize = 256
+	br := bufio.NewReader(r)
+	nodes := make([]node.Node, 0, extendedSquareSize)
+	nodeCollector := func(hash []byte, children ...[]byte) {
+		cid, err := cidFromNamespacedSha256(hash)
+		if err != nil {
+			panic(fmt.Sprintf("nmt lib created a malformed hash: %s", err))
+		}
+		switch len(children) {
+		case 1:
+			prependNode(nmtLeafNode{
+				cid:  cid,
+				Data: children[0],
+			}, &nodes)
+		case 2:
+			prependNode(nmtNode{
+				cid: cid,
+				l:   children[0],
+				r:   children[1],
+			}, &nodes)
+		default:
+			panic("expected a binary tree")
+		}
+	}
+	n := nmt.New(
+		sha256.New(),
+		nmt.NamespaceIDSize(namespaceSize),
+		nmt.NodeVisitor(nodeCollector),
+	)
+	for {
+		share := make([]byte, shareSize+namespaceSize)
+		if _, err := io.ReadFull(br, share); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if err := n.Push(share[:namespaceSize], share[namespaceSize:]); err != nil {
+			return nil, err
+		}
+	}
+	// to trigger the collection of nodes:
+	_ = n.Root()
+	return nodes, nil
+}
+
+func prependNode(newNode node.Node, nodes *[]node.Node) {
+	*nodes = append(*nodes, node.Node(nil))
+	copy((*nodes)[1:], *nodes)
+	(*nodes)[0] = newNode
+}
+
+func NmtNodeParser(block blocks.Block) (node.Node, error) {
+	const (
+		// length of the domain separator for leaf and inner nodes:
+		prefixOffset = 1
+		// nmtHashSize = flagSize+sha256.Size
+		nmtHashSize = namespaceSize*2 + sha256.Size
+	)
+	var (
+		leafPrefix  = []byte{nmt.LeafPrefix}
+		innerPrefix = []byte{nmt.NodePrefix}
+	)
+	data := block.RawData()
+	if len(data) == 0 {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: nil,
+		}, nil
+	}
+	domainSeparator := data[:prefixOffset]
+	if bytes.Equal(domainSeparator, leafPrefix) {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: data[prefixOffset:],
+		}, nil
+	}
+	if bytes.Equal(domainSeparator, innerPrefix) {
+		return nmtNode{
+			cid: block.Cid(),
+			l:   data[prefixOffset : prefixOffset+nmtHashSize],
+			r:   data[prefixOffset+nmtHashSize:],
+		}, nil
+	}
+	return nil, fmt.Errorf(
+		"expected first byte of block to be either the leaf or inner node prefix: (%x, %x), got: %x)",
+		leafPrefix,
+		innerPrefix,
+		domainSeparator,
+	)
+}
+
+var _ node.Node = (*nmtNode)(nil)
+var _ node.Node = (*nmtLeafNode)(nil)
+
+type nmtNode struct {
+	// TODO(ismail): we might want to export these later
+	cid  cid.Cid
+	l, r []byte
```

Comment:
If the above comment is applied and `nmtNode` is planned to have only links, consider using `ProtoNode` located [here](https://github.com/ipfs/go-merkledag/blob/master/node.go#L21) instead. That would return you again to only one codec for the `nmtLeafNode`, remove `nmtNode`, it's parsing, and would allow you to make deeper integration with IPFS, as most of its functionality based on `ProtoNode` right now. NMT trees can then become full-featured verifiable IPFS files and Lazyledger nodes may use that to have additional tree synchronization mechanism.

Reply:
Doesn't a `ProtoNode` change the raw bytes that get hashed in the end? I would assume a ProtoNode use protobuf when returning RawData(). I've thought about it but I think that requires us to actually change the NMT lib use protobuf too; and that's sth I'd like to avoid if possible. If the NMT can just receive (and operate on the) raw and already encoded bytes that is preferable imo.

Reply:
You are right, ProtoNode forces the use of `protobuf`, and avoiding it is better than the deeper integration I mentioned. Also, considering that the end goal is to have LL's own synchronization protocol that is only based on IPFS/IPLD and not uses it directly.